### PR TITLE
ENH: Add clang-format file

### LINKED ISSUE #19363: Add a .clang-format file for pretty printing C code
## Feature
It would be useful to have a numpy C style pretty printer, especially for newbies. Clang-format looks like a good
choice that has integrations with git (checks change sets), vim, visual studio, and others. The main work would be
in implementing the style sheet, and there are already examples for [PEP7](https://gist.github.com/pganssle/0e3a5f828b4d07d79447f6ced8e7e4db).

### PR DESCRIPTION
**Pull request by charris**
Clang-format can be used to reformat C/C++ code for codestyle fixups. The `.clang-format` file here implements the NumPy codestyle document as much as possible. The initial version here is taken from @pganssle's file at 
https://gist.github.com/pganssle/0e3a5f828b4d07d79447f6ced8e7e4db, he is listed as co-author.

I've included a formatted version of convert.c for before/after comparison, it can be removed before merging.

Closes #19363.

### CODE REVIEW

File: numpy/core/src/common/array_assign.h [hunk 1]
```diff
@@ -1,6 +1,8 @@
 #ifndef _NPY_PRIVATE__ARRAY_ASSIGN_H_
 #define _NPY_PRIVATE__ARRAY_ASSIGN_H_
 
+#include "numpy/arrayobject.h"
+
```
Comment (mattip) [inline numpy/core/src/common/array_assign.h:5]: Why the extra include?

**Comment by mattip** [comment]
> Should the three reformatted files be included?

Could we do like python: add a linter and have it report errors only on code that is part of the PR (without taking a stand on the files you have already processed)?

File: numpy/core/src/common/array_assign.h (hunk 1 above)
Comment (charris) [reply numpy/core/src/common/array_assign.h:5]: > Why the extra include?

It was to make the include order safe. However, it turns out that `.clang-format` provides a mechanism for automatically grouping and ordering  includes,  so I am going to formalize our include ordering.

**Comment by charris** [comment]
> Could we do like python

Maybe, there are several integrations that I haven't explored yet. I do know that it can be run on line ranges.

File: numpy/core/src/common/array_assign.h (hunk 1 above)
Comment (mattip) [reply numpy/core/src/common/array_assign.h:5]: As far as I can tell this include is not moved, it is new to the file.

Comment (charris) [reply numpy/core/src/common/array_assign.h:5]: > this include is not moved

True, it happened to sort after the needed `numpy/*` includes, so didn't cause a problem, but that was an accident. I'm going the put the `numpy/*` includes into their own group so they always come first and then we won't need to worry.

**Comment by charris** [comment]
> add a linter and have it report errors only on code that is part of the PR 

There are several actions available in the github Marketplace, I haven't looked closely at them. Some automatically reformat files on merge, others look to reformat differences on pushes to PRs. Not sure what would be best. If I could, I'd reformat all the C files first, that would also allow working out any kinks found in the process. But I know that there is resistance to massive reformatting at the point, although I did it in the past. The template files are also likely to pose difficulties, although I haven't tried  yet. On the plus side, most of the changes seem to involve indentation and spacing around operators, and we are already pretty good on that score. Note that at this time all the binary operators get spaces, including `*`, `/`, which is annoying but can be lived with.

**Comment by charris** [comment]
Close/reopen

**Comment by charris** [comment]
I've removed the example so that future changes after reformatting can go into separate PRs.

**Comment by pganssle** [comment]
FYI I don't recall how I came up with that original file, but presumably I did it myself since I felt comfortable releasing it the first time someone asked, so if that licensing question is still active you have my blessing (not sure if it even qualifies for copyright anyway).

**Pull request merged**